### PR TITLE
graphql-codegen: use near-operation-file

### PR DESCRIPTION
**Pull request by nsams**
The query/mutation/fragment name is used for generating interfaces, and will currently be placed in one central graphql.generated.ts file. Thus the name needs to be unique across the whole application.

For example, if you have ProductCategories you need a similar query:
- in the table where they are edited
- in the product form where they are assigned

Currently you need to somehow make sure you use different names in both components. This is inconvenient and doesn't scale well. Often the problem is simply ignored (if both queries match graphql-codegen won't complain - but that is not really a solution)

Re-using the same query in two components (or even creating a central queries file) is not a good idea, as it breaks encapsulation.

With this change we get a generated "near" the file where the query is written in. And the names don't have to be unique anymore.

Downsides:
- many generated files that can cause clutter
- not possible anymore to access `namedOperations.Query.Foo` for queries made by other components (alternatives: use a string, or export and import the query)

(In this PR I changed just demo/admin, of course we should change that "everywhere" if we decide on it)

### CODE REVIEW

File: demo/admin/src/news/Form.tsx [hunk 1]
```diff
@@ -18,11 +18,13 @@ import { ArrowLeft } from "@comet/admin-icons";
 import { EditPageLayout } from "@comet/cms-admin";
 import { IconButton } from "@mui/material";
 import { useContentScope } from "@src/common/ContentScopeProvider";
-import { GQLMutationcreateNewsArgs, GQLMutationupdateNewsArgs, GQLNewsInput, GQLNewsQuery, GQLNewsQueryVariables } from "@src/graphql.generated";
+import { GQLMutationcreateNewsArgs, GQLMutationupdateNewsArgs, GQLNewsInput } from "@src/graphql.generated";
```
Comment (nsams) [inline demo/admin/src/news/Form.tsx:21]: input types and enums stay in central file (as they are the same across the whole application)

File: demo/admin/src/pages/Page.tsx [hunk 2]
```diff
@@ -4,7 +4,8 @@ import { DocumentInterface, rewriteInternalLinks } from "@comet/cms-admin";
 import { PageTreePage } from "@comet/cms-admin/lib/pages/pageTree/usePageTree";
 import { Chip } from "@mui/material";
 import { SeoBlock } from "@src/common/blocks/SeoBlock";
-import { GQLPage, GQLPageInput, GQLPageTreeNodeAdditionalFieldsFragment } from "@src/graphql.generated";
+import { GQLPageTreeNodeAdditionalFieldsFragment } from "@src/common/EditPageNode";
```
Comment (nsams) [inline demo/admin/src/pages/Page.tsx:7]: this import was invalid before:

PageTreeNodeAdditionalFields Fragment is defined in EditPageNode.tsx and additionally used in Page.tsx. This is (strictly speaking) invalid, as the fragment is a private part of EditPageNode.tsx. Easy fix: make it public by exporting it, then we can import it here.

File: demo/admin/src/App.tsx [hunk 3]
```diff
@@ -47,6 +47,7 @@ import { NewsLinkBlock } from "./news/blocks/NewsLinkBlock";
 import News from "./news/News";
 import MainMenu from "./pages/mainMenu/MainMenu";
 import { Page } from "./pages/Page";
+import ProductCategoriesPage from "./products/categories/ProductCategoriesPage";
```
Comment (fraxachun) [inline demo/admin/src/App.tsx:50]: Target file is missing

File: demo/admin/src/common/EditPageNode.tsx [hunk 4]
```diff
@@ -4,6 +4,7 @@ import { createEditPageNode } from "@comet/cms-admin";
 import { MenuItem } from "@mui/material";
 import * as React from "react";
 import { FormattedMessage } from "react-intl";
+export { GQLPageTreeNodeAdditionalFieldsFragment } from "./EditPageNode.generated"; //re-export
```
Comment (fraxachun) [inline demo/admin/src/common/EditPageNode.tsx:7]: Target generated file has a wrong import (at least on my tests):
`import * as Types from '../src/graphql.generated';`
It should be
`import * as Types from '../../src/graphql.generated';`

**Review by fraxachun — CHANGES_REQUESTED** [review]

File: demo/admin/src/common/EditPageNode.tsx (hunk 4 above)
Comment (fraxachun) [reply demo/admin/src/common/EditPageNode.tsx:7]: This is still wrong...

Comment (nsams) [reply demo/admin/src/common/EditPageNode.tsx:7]: somehow forgot to git add a file :/

**Comment by nsams** [comment]
> In general I like this very much and consider it as an improvement. However, would it be possible to let prettier format the generated files automatically?

yes, however it comes with a cost (time):

- no formatter: 1.9s
- prettier: 2,7s
- eslint: 4.7s

eslint does additional (useful) work like removing unused imports and sorting imports.

Possible alternative: ignore generated files in eslint (we did that for the single generated file) and you won't see the errors in the IDE

**Comment by fraxachun** [comment]
I get the following error with the re-exports but don't know why and I'm not able to resolve it (I don't see an obvious error):
`export 'GQLPageTreeNodeAdditionalFieldsFragment' (reexported as 'GQLPageTreeNodeAdditionalFieldsFragment') was not found in './EditPageNode.generated' (possible exports: namedOperations)`

Why exactly shouldn't it be allowed to import directly from generated files?

**Comment by nsams** [comment]
> Does this affect the watcher to the same degree? Every change needs 2.8s more?

I guess so

**Comment by nsams** [comment]
> I get the following error with the re-exports but don't know why and I'm not able to resolve it (I don't see an obvious error):

fixed. (expot type is required)

> Why exactly shouldn't it be allowed to import directly from generated files?

EditMainMenuItem fragment is defined in EditMainMenuItem.tsx and only this file must use it as it's a private part of this file. This was not really obvious before near-operation-file but now it is. Should be checked using https://github.com/vivid-planet/comet/pull/1096.

File: demo/admin/package.json [hunk 5]
```diff
@@ -32,6 +32,8 @@
         "@emotion/react": "^11.9.3",
         "@emotion/styled": "^11.9.3",
         "@fontsource/roboto": "^4.5.5",
+        "@graphql-codegen/near-operation-file-preset": "^2.5.0",
+        "@graphql-codegen/typescript-react-apollo": "^3.3.7",
```
Comment (johnnyomair) [inline demo/admin/package.json:36]: Please remove

**Review by dkarnutsch — DISMISSED** [review]

**Review by manuelblum — APPROVED** [review]

**Comment by nsams** [comment]
@fraxachun I failed to enable eslint and left if off for now (not directly to this PR related change)

Problem was with ignore patterns (we need to ignore generated files during "normal" eslint runs but not during codegen runs)

File: demo/admin/package.json [hunk 6]
```diff
@@ -32,6 +32,7 @@
         "@emotion/react": "^11.9.3",
         "@emotion/styled": "^11.9.3",
         "@fontsource/roboto": "^4.5.5",
+        "@graphql-codegen/near-operation-file-preset": "^2.5.0",
```
Comment (johnnyomair) [inline demo/admin/package.json:35]: All the other graphql-codegen packages are in dev dependencies, please move it there

File: demo/admin/src/pages/mainMenu/components/EditMainMenuItem.tsx [hunk 7]
```diff
@@ -109,7 +106,7 @@ const EditMainMenuItem: React.FunctionComponent<EditMainMenuItemProps> = ({ item
                 input: { content: content ? RichTextBlock.state2Output(content) : null },
                 lastUpdatedAt: item.updatedAt,
             },
-            refetchQueries: [namedOperations.Query.MainMenuItem],
```
Comment (johnnyomair) [inline demo/admin/src/pages/mainMenu/components/EditMainMenuItem.tsx:112]: Why doesn't named operations work anymore?

File: demo/site/preBuild/src/createRedirects.generated.ts [hunk 8]
```diff
@@ -0,0 +1,14 @@
+import * as Types from '../../src/graphql.generated';
```
Comment (johnnyomair) [inline demo/site/preBuild/src/createRedirects.generated.ts:1]: Was this committed intentionally?

File: demo/site/preBuild/src/publicGenerator/sitemap.xml.generated.ts [hunk 9]
```diff
@@ -0,0 +1,14 @@
+import * as Types from '../../../src/graphql.generated';
```
Comment (johnnyomair) [inline demo/site/preBuild/src/publicGenerator/sitemap.xml.generated.ts:1]: Same as above

File: demo/site/src/components/Breadcrumbs.generated.ts [hunk 10]
```diff
@@ -0,0 +1,9 @@
+import * as Types from '../graphql.generated';
```
Comment (johnnyomair) [inline demo/site/src/components/Breadcrumbs.generated.ts:1]: Did you forget to add a .gitignore in site? 😁

File: packages/admin/cms-admin/src/blocks/DamVideoBlock.generated.ts [hunk 11]
```diff
@@ -0,0 +1,13 @@
+import * as Types from '../graphql.generated';
```
Comment (johnnyomair) [inline packages/admin/cms-admin/src/blocks/DamVideoBlock.generated.ts:1]: Here as well?

File: packages/admin/cms-admin/src/builds/BuildEntry.generated.ts [hunk 12]
```diff
@@ -0,0 +1,11 @@
+import * as Types from '../graphql.generated';
+
+export const namedOperations = {
```
Comment (johnnyomair) [inline packages/admin/cms-admin/src/builds/BuildEntry.generated.ts:3]: Here the named operations still exist 🤔

**Comment by johnnyomair** [comment]
Please use the generated named operations wherever possible (might need to reexport them as well)

File: demo/admin/src/pages/mainMenu/components/EditMainMenuItem.tsx (hunk 7 above)
Comment (nsams) [reply demo/admin/src/pages/mainMenu/components/EditMainMenuItem.tsx:112]: because MainMenuItem Query is defined in another file

**Review by johnnyomair — APPROVED** [review]

**Review by johnnyomair — APPROVED** [review]

**Review by fraxachun — APPROVED** [review]

**Pull request merged**